File: 2020/day2/part_one.go
package main

import (
	"bufio"
	"log"
	"os"
	"strconv"
	"strings"
)

func main() {

	file, err := os.Open("input.txt")
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	cnt := 0

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {

		line := scanner.Text()
		parts := strings.Split(line, " ")
		if len(parts) != 3 {
			log.Fatalf("error on line with content %s", line)
		}

		timeParts := strings.Split(parts[0], "-")
		minTimes, err := strconv.Atoi(string(timeParts[0]))
		if err != nil {
			log.Fatal(err)
		}
		maxTimes, err := strconv.Atoi(string(timeParts[1]))
		if err != nil {
			log.Fatal(err)
		}

		policyChar := rune(parts[1][0])

		charCount := 0
		for _, letter := range parts[2] {
			if letter == policyChar {
				charCount++
			}
		}

		if charCount >= minTimes && charCount <= maxTimes {
			cnt++
		}
	}

	log.Print(cnt)
}

File: 2021/day03/part_one.py
import math


def main():

    bits = {}

    with open("input.txt") as f:

        for line in f:
            line = line.strip()

            for i in range(len(line)):
                if i not in bits:
                    bits[i] = 0

                if line[i] == "0":
                    bits[i] -= 1
                else:
                    bits[i] += 1

    gamma, epsilon = 0, 0
    for i in range(len(bits)):
        if bits[i] > 0:
            gamma += math.pow(2, len(bits)-1-i)
        else:
            epsilon += math.pow(2, len(bits)-1-i)

    print(gamma * epsilon)


if __name__ == "__main__":
    main()
